### DISCUSSION TITLE
Discussion on migration to Neovim 0.9 and major refactor

### DISCUSSION THREAD

**Discussion by ful1e5** [projekt0n/github-nvim-theme#226] [General]
**[Update 1](https://github.com/projekt0n/github-nvim-theme/discussions/226#discussioncomment-5266253)**

TLDR; I'm planning to refactor the codebase for compatibility with Neovim 0.9 and is seeking feedback on requested features and improvements.

I have been unable to push new features for the past few months due to bugs and limitations with the codebase for this color scheme. The codebase was heavily inspired by [tokyonight.nvim](https://github.com/folke/tokyonight.nvim), which was the first color scheme written in pure Lua and had many features for Neovim 0.5, including support for tree-sitter highlights. However, Neovim has exposed more of vim's native commands through `vim.*`, which has made many of the utility functions we wrote no longer useful.

Additionally, Treesitter recently removed the `TS*` highlight group, which we patched in our codebase (PR #222). However, we can do a better job by reducing common highlights according to the new variant of the GitHub color scheme that we plan to implement. That is why I am planning to reorganize the code to use Neovim 0.9's features and native commands in a major refactor.

In order to **minimize the impact on people's Neovim configurations**, I would like to issue a warning before the refactor begins. If any major changes are pushed upstream, I will provide migration documentation for those following the main branch. By starting this discussion beforehand and sharing the changes and problems I have identified, we can address the most requested features, make the code extensible for future requests, and determine the best way to make these improvements.

By making these changes, we can not only make it easier for new people to join this community and contribute to the project, but we can also enable others to use this color scheme as a template for creating their own. By sharing the changes and problems I have found in detail, we can discuss them and determine the best way to make these improvements.

## Addressing Key Limitations

Here are the improvements we are making to this colorscheme:

- [Restructuring Modules](#restructuring-modules)
- [Ensuring Future Proof Configuration](#ensuring-future-proof-configuration)
- [Adhering to primer for Color Guide](#adhering-to-primer-for-color-guide)
- [Removing Aesthetic Elements Inspired by VSCode Themes](#removing-aesthetic-elements-inspired-by-vscode-themes)
- [Implementing Requested Features](#implementing-requested-features)

### Restructuring Modules

The current structure of the project's modules presents challenges that can make it difficult for new contributors to understand and contribute to the project, as well as for existing contributors to read and refactor the tightly-coupled functions and modules.

One significant issue is the use of inconsistent naming conventions and arbitrary divisions of functions, which can make it difficult to understand the project's structure. Another problem is the lack of clear comments or documentation, which could greatly assist someone in working on the project.

These issues were exacerbated when additional modules were added to support native autocmd and support for old vim syntaxes, as well as lua type support and new feature configuration options. Given these challenges, it would be beneficial to reorganize and clarify the project's structure.

The current module structure is as follows:

```
<project root>
├─ colors                             <-- This special directory is responsible for exposing the colorscheme to Neovim and changing the colorscheme with the `:colo` command.
│  ├─ github_dark.vim
│  ├─ github_dark_colorblind.vim
│  ├─ github_dark_default.vim
│  ├─ github_dimmed.vim
│  ├─ github_light.vim
│  ├─ github_light_colorblind.vim
│  └─ github_light_default.vim
├─ doc                                <-- Exposes the documentation with the `:h` command.
│  └─ gt_changelog.txt                <-- This file contains detailed changelogs and is exposed to `:h github-theme.changelog`.
└─ lua
   ├─ github-theme
   │  ├─ config
   │  │  ├─ default.lua               <-- This is the default configuration file for the colorscheme.
   │  │  └─ vim_config.lua            <-- This converts the `vim.g.github.*` values to Lua values.
   │  ├─ palette                      <-- This module holds the hex values for the colors in the colorscheme for the different variants.
   │  │  ├─ dark.lua
   │  │  ├─ dark_colorblind.lua
   │  │  ├─ dark_default.lua
   │  │  ├─ dimmed.lua
   │  │  ├─ light.lua
   │  │  ├─ light_colorblind.lua
   │  │  └─ light_default.lua
   │  ├─ plugins                      <-- This module is currently being refactored to have highlights for individual files for plugins, but the process has been halted due to larger issues.
   │  │  └─ lualine.lua
   │  ├─ types                        <-- This adds type support in the Lua documentation and is also useful for type checking.
   │  │  ├─ gt.lua
   │  │  ├─ init.lua
   │  │  └─ palette.lua
   │  ├─ autocmds.lua                 <-- Responsible for defining the autocmds for the colorscheme.
   │  ├─ colors.lua                   <-- Provides an extra layer for filtering colors by variant, using the palette files, and also defines common colors.
   │  ├─ config.lua                   <-- Used to override user configurations and also processes the user's Vim configuration.
   │  ├─ init.lua                     <-- Exposes the `setup` function.
   │  ├─ palette.lua                  <-- Returns the appropriate palette based on the request.
   │  ├─ theme.lua                    <-- Holds all of the highlight values.
   │  └─ util.lua                     <-- This module contains miscellaneous functions and is where much of the processing occurs.
   └─ lualine
      └─ themes                       <-- This module exposes the github themes within the lualine configuration.
         ├─ github_dark.lua
         ├─ github_dark_colorblind.lua
         ├─ github_dark_default.lua
         ├─ github_dimmed.lua
         ├─ github_light.lua
         ├─ github_light_colorblind.lua
         └─ github_light_default.lua
```

I am describing the structure because it was requested in #193, and I believe it may be helpful for people to have a better understanding of the changes being made.

As we continue to work on optimizing the project, **it is important to me that we maintain the current features provided by the colorscheme**. To ensure a smooth refactoring process, I am proposing the use of built-in commands and the revision of certain functions. **If you have any thoughts on the project's overall structure, I would be grateful for your input.**

For ideas on function naming and module organization, I want to refer to the [ellisonleao/gruvbox.nvim](https://github.com/ellisonleao/gruvbox.nvim) colorscheme. **If you are a plugin or colorscheme author, please feel free to share your ideas on structure as well.**

In terms of specific areas of focus, I plan to reorganize the `util.lua` module into a utility module. For the `theme.lua` module, I am considering grouping plugin highlights in the form of a Lua table, although I am not yet certain what direction to take with this module. Finally, I also plan to perform a refactor on the `config.lua` module to make it more adaptable to future feature implementations, which we can discuss further below. Suggestions for renaming these module names to improve project readability are also welcome.

### Ensuring Future Proof Configuration

We have been gradually adding new configurations to our project. Currently, when a new feature or option is introduced, we add it to the Lua config table and name it appropriately. However, I think it would be more effective to have a more organized and consistent config structure. By using Lua tables to group and understand the configurations, we can make the configuration easier to understand and navigate without needing to refer to documentation.

I would like to create an extendable config module that will make it easier to add new features in the future. To do this, I plan to carefully review the config structures of other colorschemes and group similar configs together in Lua tables. I am also planning to provide clear typing and detailed documentation for each option. One project that I find particularly interesting in this regard is [navarasu/onedark.nvim](https://github.com/navarasu/onedark.nvim). I welcome any suggestions or ideas you may have on this topic.

### Adhering to primer for Color Guide

I am also considering renaming the colorscheme variant to align with the guidelines provided by [primer.style](https://primer.style/primitives/colors) and the colorblind variant must be renamed because it was removed from the beta version on primer website. It is uncertain if the variant of the VSCode theme is currently following these guidelines, but I believe it is being developed in accordance with them. To ensure consistency with other GitHub products and prevent potential issues, it is important to follow the guidelines provided by primer.

> **Note**
> Changes are highlighted in bold text.

| Current Name             | VSCode Name                    | Github Appearance Setting       | Primer Name             | New Name                                                                |
| ------------------------ | ------------------------------ | ------------------------------- | ----------------------- | ----------------------------------------------------------------------- |
| github_dark              | GitHub Dark                    | Dark default                    | Dark                    | github_dark                                                             |
| **github_dimmed**        | **GitHub Dark Dimmed**         | **Dark dimmed**                 | **Dark dimmed**         | **github_dark_dimmed**                                                  |
| **github_dark_default**  | **GitHub Dark Default**        | **N/A**                         | **N/A**                 | **"Planning to remove legacy code and switch to dark_high_contrast."**  |
| **N/A**                  | **GitHub Dark High Contrast**  | **Dark high contrast**          | **Dark high contrast**  | **github_dark_high_contrast**                                           |
| github_dark_colorblind   | GitHub Dark Colorblind (Beta)  | Dark Protanopia & Deuteranopia  | Dark colorblind         | github_dark_colorblind                                                  |
| **N/A**                  | **N/A**                        | **Dark Tritanopia**             | **Dark tritanopia**     | **github_dark_tritanopia**                                              |
| github_light             | GitHub Light                   | Light default                   | Light                   | github_light                                                            |
| **github_light_default** | **GitHub Light Default**       | **N/A**                         | **N/A**                 | **"Planning to remove legacy code and switch to light_high_contrast."** |
| **N/A**                  | **GitHub Light High Contrast** | **Light high contrast**         | **Light high contrast** | **github_light_high_contrast**                                          |
| github_light_colorblind  | GitHub Light Colorblind (Beta) | Light Protanopia & Deuteranopia | Light colorblind        | github_light_colorblind                                                 |
| **N/A**                  | **N/A**                        | **Light Tritanopia**            | **Light tritanopia**    | **github_light_tritanopia**                                             |

The current color palette used in this colorscheme was selected manually using a color picker, which may cause some differences in value of the colors compared to the palette used by VSCode themes and primer.

I also found that VSCode themes do not use a universal highlighter for code highlighting, but rather rely on community-created plugins using their preferred highlighter, which may involve some stupid regex stuff. This can cause certain types of code to be unparsed and result as default foreground color. This issue was not previously recognized when defining highlights in the colorscheme. Therefore, it is being considered to fix this issue in this refactor without significantly changing the VSCode and Neovim themes.

In the event that VSCode implements treesitter as a highlighter in the future, the highlights will be revisited and potentially modified through discussion.

### Removing Aesthetic Elements Inspired by VSCode Themes

It is not necessary to completely remove highlights inspired by VSCode themes, but if a particular highlight (such as `@string.special`) is connected to a core highlight and is part of a chain (for example, `Special->@string.special->@lua.string.special`), replacing it will cause all child highlights to become disconnected, and you will have to manually define all highlights linked to it (such as `@string.special`). This can be inconvenient and time-consuming.

Additionally, highlights for Treesitter and Treesitter-based plugins often change, so you may need to constantly update these highlights as well. However, we can potentially save time by only overriding core Treesitter highlights and highlights related to supported programming languages.

### Implementing Requested Features

I would like to implement feature request #176 as part of this refactor. Additionally, I would like to add documentation about the ability to override highlighting for specific file types, as discussed in #213. While requests for plugin support will not be addressed during the refactor, we should plan to address them after the refactor is complete.

I am grateful for everyone's efforts in reading my lengthy list of refactor suggestions and participating in the discussion. Your support and sponsorship have not gone unnoticed and I would like to especially thank @github for being my first one-time sponsor. Thank you all for your invaluable contributions and support.

**Comment by adrian5** [comment]
Another tidbit of information: The **Nightfox** colorscheme has a [compiler](https://github.com/EdenEast/nightfox.nvim/#compile) feature. If this could lay to rest any need for tradeoffs between an easy to maintain (modular) design and colorscheme performance, it might be worth looking into.

Maybe nvim 0.9 will already do other caching optimizations in this regard, I don't know.

**Reply by ful1e5** [reply]
This is exactly what I was hoping for. I have been pondering on the startup time of the colorscheme and the possibility of toggling highlighting for individual plugins that aren't installed on the system. I believe compiling the colorscheme would be a suitable solution. I would like to incorporate this feature into a refactor. Thank you for the suggestion.

**Reply by ful1e5** [reply]
@adrian5 Implemented in 2c782a36abd5d5b541891f2a2309514cc394f37c . Give it a try. :+1:

**Comment by ful1e5** [comment]
**Update 1**

I apologize for the delay in starting the refactoring project as I have been preoccupied with other tasks. However, I plan to begin the project soon and will be streaming the process on [twitch.tv/ful1e5](http://twitch.tv/ful1e5).

Additionally, I would like to mention that neovim has recently introduced a breaking change with semantic highlight groups. As a result, I will be incorporating this into the project's restoration. The new highlight group names will be under `@lsp.*` and will be specific to each file type. For more information about this change, you refer to `:h lsp-semantic-highlight` on neovim nightly or track https://github.com/neovim/neovim/commit/1cc23e1109ed88275df5c986c352f73b99a0301c.

**Comment by tmillr** [comment]
Parens and square brackets are still the wrong color. They should be the default foreground color.

<img width="795" alt="Screenshot 2023-05-06 at 1 53 41 AM" src="https://user-images.githubusercontent.com/45028928/236614070-416b1350-309a-4fe0-8e2a-07551cc08b8b.png">

I'm using github dark dimmed and am on the latest commit as of this post.

Also, I believe table properties are supposed to be fg color too (e.g. `callback` in the screenshot above), at least when it comes to lua.

Here is how it is supposed to look:

<img width="438" alt="Screenshot 2023-05-06 at 1 58 56 AM" src="https://user-images.githubusercontent.com/45028928/236614336-215745c5-f6bd-48b7-9742-603f90292be9.png">

And I could be wrong, but I think diffAdded and diffRemoved etc. should have the default foreground color as their foreground setting as well. This is what gitsigns uses by default to hl line numbers and the numbers are ending up green (their foreground setting should be the same color as the default/global foreground color)...but their bg setting seems to be correct at least (i.e. green and red respectively).

**Reply by tmillr** [reply]
## How it is supposed to be

In some langs (like js) method and regular fn calls are indeed pink (relative to the GitHub dark dimmed theme), in lua they are blue as shown in the screenshot.

For Lua, function names in function definitions/declarations are pink. `self` (and probably all other builtins vars, including `nil`) is blue like a fn call. Booleans are the same blue, as are numbers. Fn parameters part of funcdef are unhighlighted or default fg color, including `...` it seems. `.` property access names are also unhighlighted, as in `xxx.unhighlighted`. 

Regardless of lang, parens and braces should always be the default fg color (or simply unhighlighted) from what I've seen. The only exception to this in Lua are the square brackets used to delimit multiline comments (in which case they should be highlighted the same as comment text), or those used to delimit strings (in which case they should be light blue like the string content).

```lua
--[[
lua
--]]

-- builtin fn
print()
table.insert()
builtin = table.insert
concat = "" .. "" .. ""

local CAPITALIZED = 2 + 2 / 2 and true

-- ... is default fg color here and so are ( and ) and params
function pink(a, b, ...)

end

local function pink()

end

-- . is pink here
function pink.pink()

end

function a.b.c()
end

-- : is pink here
function pink:pink()
    self:blue() -- self is blue too
    self.blue()
    self.x = y
    y = self.x
    y = self
    self.x = debug
    self.x = debug.debug
    self.x = table
end

function a.b:c()
end

local default_fg = 9999.99
local d, d2 = debug, debug.setmetatable

-- All string delimiters including their content are light blue
local default_fg = [[light blue]], 'light blue', "light blue"

Blue { default_fg = nil }
Blue()
default_fg.AlsoBlue {}
default_fg:AlsoBlue {}
default_fg.AlsoBlue()
default_fg:AlsoBlue()
local default_fg
default_fg
default_fg = default_fg.default_fg
local default_fg = { default_fg = true }
-- etc.

-- unless otherwise noted , and . and : and [ and ] etc. are default fg color...basically all punct chars it seems
```

### JSON

```json
{ "key": "val\nue", "arr": [1, 2, 3, true, null] }
```

### JS

```js
class MyType {
  constructor(param) {
    super()
    this.abc = this
    super.method(undefined, null)
  }

  method(param, ...k) {}
}


// builtin
setTimeout(() => {})
console.log()

const b = MyType();
const b = new MyType();
let regex = /^123([abc])+\1\2(?:jddu)$/g;
var str = `${exp + 2}`
const x = y ? true : false;

function func(a, b, ...c) {

}

func(1, a, { x: y }, 2 + 2)

const obj = { method(a, b, ...c) {}, prop: true }

const x = obj.prop.prop
const x = obj.prop.prop()
const x = obj.prop()
debugger
```

### TS

```ts
type SomeType = "" | number

class MyType {
  constructor(param: string) {
    super()
    this.abc = this
    super.method(undefined, null)
  }

  method(p: MyType) {}
}

// builtin
setTimeout(() => {})
console.log()
const b = MyType();
const b = new MyType();
let regex = /^123([abc])+\1\2(?:jddu)$/g;
var str = `${exp + 2}`
const x = y ? true : false;

function func(a, b, ...c) {

}

func(1, a, { x: y }, 2 + 2)

const obj = { method(a, b, ...c) {}, prop: true }

const x = obj.prop.prop
const x = obj.prop.prop()
const x = obj.prop()
debugger
```

### Ruby

```ruby
hash = { "Jane Doe" => 10, "Jim Doe" => 6 }
hash = { aa: 10, bb: 6 }

class Cls < Sub
  attr_accessor :abc
  @one = "abc"
  @@two = "def"
end

def some_method(value='default', arr=[])
  puts value
  puts arr.length
end

results = method_name parameter1, parameter2
results = method_name(parameter1, parameter2).reverse
```

### Make

```make
SHELL = bash
.PHONY: target target2
v = $(call myfn,arg)
override somevar = hello
xxx := hello
yyy ::= hello

override define multi
  @goodbye $$world "$$(echo world)"
endef

target2: ;
target: preq

```

### Rust

```rust
#![crate_type = "lib"]
use allmine::stillmine::Func;
use std::Vec;

pub MyStruct {
    pub id: Option<usize>,
    hello: MyStruct,
}

// This is a simple macro named `say_hello`.
macro_rules! say_hello {
    // `()` indicates that the macro takes no argument.
    () => {
        // The macro will expand into the contents of this block.
        println!("Hello!");
    };
}

#[allow(non_camel_case_types)]
#[test]
#[derive(Clone, Debug)]
fn somefunc<'a>(param: Option<bool>, p2: SomeType<'a>, p3: Self, p4: Self::SomeType) -> int32 {
    let user1 = User {
        hello: mod::ns::ok,
        active: true,
        username: String::from("someusername123"),
        email: String::from("someone@example.com"),
        sign_in_count: 1,
    };

    println!("Hello!");
    let abc = 100 + 100u8;
    let abc = user1.active;
    let xyz = Some(User2 { active: [ 1, 2, 3 ] });
    let f = NotSome(None);
    Abc::new();
    self.meth()
}
```

### HTML

```html
<html>
  <div class="someclass">
    Hello World!
  </div>
<html>
```

### CSS

```css
div.div#foo {
  color: rgba(100, 200, 0, 0.5) !important;
}
```

### C

```c
#include <stdio.h>
#include <string.h>
 
struct Books {
   char  title[50];
   char  author[50];
   char  subject[100];
   int   book_id;
};

main() {
  unsigned int myint = 42;
  struct Books Book1;
  struct Books Book2;
 
  strcpy( Book1.title, "C Programming");
  somefunc();
  Book1.book_id = &book;
  Book1.book_id = *book;
  Book1->book_id = 'c';

  return 0;
}
```

```bash
command mycmd ${var} ${#var} "$var" "${var}" | cat
true || false

somecmd ${var} ${#var} "$var" "${var}" $var
IFS=$IFS
ifs=$ifs
var=$(command somecmd "$(echo hello)" $(mycmd "hello" hello) "$(mycmd "hello" hello)")

func() {
  :
  return 0
}

func 11 ${arr[@]} ${arr[*]} "${arr[@]}" "${arr[*]}" $IFS $var $(<file.lua) < somefile.lua

cmd \
    -p1 \
    --p2 \
    p3

cmd1 p1; cmd2;

exit
```

**Reply by tmillr** [reply]
Can we change this? https://github.com/projekt0n/github-nvim-theme/blob/3d62e7b679f15940ecd5cd721d50d1c692853149/lua/github-theme/palette/github_dark_dimmed.lua#L187

**Reply by ful1e5** [reply]
@tmillr You can override `specs` in the `setup` of a theme, like this:

```lua
local colo = 'github_dark'
local spec = require('github-theme.spec').load(colo)
local palette = spec.palette  -- To access palette values

require('github-theme').setup({
  specs = {
    all = {
      syntax = {
        bracket = spec.fg1,
      },
    },
  },
})

vim.cmd.colorscheme(colo)
```

**Reply by tmillr** [reply]
@ful1e5 Alright I'll check that out. Thank you

**Reply by ful1e5** [reply]
@tmillr If you are trying to solve the issue of highlight mismatch and would like me to review your pull request based on the description above, I would be happy to do so. In the meantime, I will be working on other features.

**Comment by tmillr** [comment]
`:helptags` on this repo fails with `E154: Duplicate tag "'colors'" in file ../.local/share/nvim/plugged/github-nvim-theme/doc/gt_deprecated.txt`

there are 2 `'colors'` tags in gt_deprecated.txt

**Reply by ful1e5** [reply]
@tmillr Track #246

**Reply by ful1e5** [reply]
Fixed in 9ddfac19923b7a615b29215194f7dd66fd4a431a

**Comment by tmillr** [comment]
Also, this isn't working for me. Am I doing something wrong?

```lua
    groups = {
        all = {
            -- this will remove the highlight groups
            ["@punctuation.bracket"] = {},
        },
```

**Reply by ful1e5** [reply]
Fixed in 861245ed4be1356c59b7d87ee55057e6d37f65e9

**Comment by borisrorsvort** [comment]
White theme (GitHub_light) colours seem really different from the actual GitHub theme (on the right). Contrast, highlight groups, and colour assignments are way better on GitHub currently.
<img width="1452" alt="Capture d’écran 2023-05-09 à 10 45 40" src="https://user-images.githubusercontent.com/39081/237044357-9d8c6cbe-ab5c-4939-98c3-63a174fa92af.png">

**Reply by borisrorsvort** [reply]
And yes, it’s the latest code from the main branch

**Reply by ful1e5** [reply]
@borisrorsvort Which language is this, and does this happen with all light variants?

**Reply by tmillr** [reply]
> @borisrorsvort Which language is this, and does this happen with all light variants?

I think it's ruby.

**Reply by borisrorsvort** [reply]
yep it’s ruby

**Reply by tmillr** [reply]
The colors are wrong, but the washed-out appearance could just be your font settings (or the way your terminal is rendering text). What terminal or gui are you using? Maybe try messing with anti-aliasing settings or increasing the font weight of your regular/default font within the settings of the terminal you're using.

**Reply by borisrorsvort** [reply]
I’m using Iterm (though the iterm colour scheme is broken and cannot be imported so im using a default light theme) with Fira code nerd font. I just saw there was a retina weight and that does improve the washed out effect (thanks for the tip).

About the colours, is this something on your list already?

**Reply by tmillr** [reply]
> About the colours, is this something on your list already?

Yeah. I have some ideas and changes that I'd like to contribute and run past the owner/maintainer concerning the incorrect highlights, and It sounds like this will become a priority soon. I believe that it all starts with #253. You can also check out #252 and #237.

**Comment by tmillr** [comment]
Just a suggestion, but it might be a good idea to note in the documentation and/or README (if not done already) that this plugin breaks the links from `@*` to `TS*` hl groups. This isn't necessarily a bad thing, but it does mean that the user needs to use the new/native `@*` treesitter hl groups (aka treesitter captures) instead of the `TS*` ones when overriding them. I'm not totally sure where the `TS*` groups come from, either Neovim itself or the [nvim-treesitter](https://github.com/nvim-treesitter/nvim-treesitter) plugin.

Edit: some of the links back to vim's standard/default hl groups are broken too, for example the link from `@variable` to `Identifier`. These links should not be broken unless a specific language's syntax highlighting has a different color from the default, in which case `@variable.LANG` should be set, not `@variable` itself (which will break the aforementioned link to `Identifier`). Keeping the links intact also makes it easier on the user when they want to override a group for multiple languages (e.g. `Identifier`, which should be the hl group for `@variable`, and should probably be set to whatever is the most common identifier color among all the languages). The default links can be found via `:help treesitter-highlight-groups`.

Just my 2 cents

**Comment by tmillr** [comment]
I submitted a pr #250 to fix lua function declarations (the highlighting of the function's/method's name).

**Comment by ful1e5** [comment]
**A major version, [v1.0.0](https://github.com/projekt0n/github-nvim-theme/releases/tag/v1.0.0), has been released.** I would like to express my gratitude to everyone who contributed to and sponsored this project. A special thank you goes to @tmillr for enhancing the highlights and providing additional support for missing Treesitter's highlights.

I'm concluding the discussion now. If anyone needs assistance, please start a new conversation. If you encounter any issues, feel free to create an issue.

Finally, I kindly request you to consider sponsoring my work to support its continuous development. Your sponsorship would be greatly appreciated.

**Reply by adrian5** [reply]
Awesome work @ful1e5 and @tmillr!
